### ISSUE TITLE
pyright 1.1.232 cannot match protocol-based `partial` with `functools.partial` 

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#3255]
Prior to pyright 1.1.232, I was able to implement a protocol that described `functools.partial`. Now, I see the following error (running on Python 3.8):

```python
import functools

from typing_extensions import Protocol, Self
from typing import Callable, Dict, Tuple, Type, TypeVar

_T = TypeVar("_T")

class Partial(Protocol[_T]):
    @property
    def func(self) -> Callable[..., _T]:
        ...

    @property
    def args(self) -> Tuple[Any, ...]:
        ...

    @property
    def keywords(self) -> Dict[str, Any]:
        ...

    def __new__(
        cls: Type[Self], __func: Callable[..., _T], *args: Any, **kwargs: Any
    ) -> Self:
        ...

    def __call__(self, *args: Any, **kwargs: Any) -> _T:
        ...

def f(x: Partial[int]):
    ...

f(functools.partial(int))  # pyright == 1.1.232: error (see below)
```

The error message is:

```
Argument of type "partial[int]" cannot be assigned to parameter 
"x" of type "Partial[int]" in function "f"
    "partial[int]" is incompatible with protocol "Partial[int]"
      "__new__" is an incompatible type
        Type "(cls: Type[Self@__new__], __func: (...) -> int, *args: Any, **kwargs: Any) -> Self@__new__" cannot be assigned to type "(cls: Type[partial[int]], __func: (...) -> _T@Partial, *args: Any, **kwargs: Any) -> partial[int]"
          Parameter 2: type "(...) -> _T@Partial" cannot be assigned to type "(...) -> int"
            Type "(...) -> _T@partial" cannot be assigned to type "(...) -> int" 
```

**Comment by erictraut** [comment]
Thanks for the bug report. This was due to a long-standing bug, but it was exposed by a recent change that was made. It affects protocol matching when a property returns a generic type, as `func` does in your example. I've fixed the problem and added a new unit test case based on your sample.

**Comment by erictraut** [comment]
This is included in pyright 1.1.233, which I just published. It will also be included in the next release of pylance.
